impeller/renderer/backend/vulkan/driver_info_vk.cc

```diff
@@ -7,17 +7,138 @@
+    {"702", AdrenoGPU::kAdreno702},
+
+    // 600
+    {"695", AdrenoGPU::kAdreno695},
+    {"690", AdrenoGPU::kAdreno690},
+    {"685", AdrenoGPU::kAdreno685},
+    {"680", AdrenoGPU::kAdreno680},
+    {"675", AdrenoGPU::kAdreno675},
+    {"663", AdrenoGPU::kAdreno663},
+    {"660", AdrenoGPU::kAdreno660},
+    {"650", AdrenoGPU::kAdreno650},
+    {"644", AdrenoGPU::kAdreno644},
+    {"643L", AdrenoGPU::kAdreno643L},
+    {"642", AdrenoGPU::kAdreno642},
+    {"642L", AdrenoGPU::kAdreno642L},
+    {"640", AdrenoGPU::kAdreno640},
+    {"630", AdrenoGPU::kAdreno630},
+    {"620", AdrenoGPU::kAdreno620},
+    {"619", AdrenoGPU::kAdreno619},
+    {"619L", AdrenoGPU::kAdreno619L},
+    {"618", AdrenoGPU::kAdreno618},
+    {"616", AdrenoGPU::kAdreno616},
+    {"615", AdrenoGPU::kAdreno615},
+    {"613", AdrenoGPU::kAdreno613},
+    {"612", AdrenoGPU::kAdreno612},
+    {"610", AdrenoGPU::kAdreno610},
+    {"608", AdrenoGPU::kAdreno608},
+    {"605", AdrenoGPU::kAdreno605},
+    // 500
+    {"540", AdrenoGPU::kAdreno540},
+    {"530", AdrenoGPU::kAdreno530},
+    {"512", AdrenoGPU::kAdreno512},
+    {"510", AdrenoGPU::kAdreno510},
+    {"509", AdrenoGPU::kAdreno509},
+    {"508", AdrenoGPU::kAdreno508},
+    {"506", AdrenoGPU::kAdreno506},
+    {"505", AdrenoGPU::kAdreno505},
+    {"504", AdrenoGPU::kAdreno504},
+};
+
+const std::unordered_map<std::string_view, MaliGPU> kMaliVersions = {
+    // 5th Gen.
+    {"G925", MaliGPU::kG925},
+    {"G725", MaliGPU::kG725},
+    {"G625", MaliGPU::kG625},
+    {"G720", MaliGPU::kG720},
+    {"G620", MaliGPU::kG620},
+
+    // Valhall
+    // Note: there is an Immortalis-G715 a Mali-G715
+    {"G715", MaliGPU::kG715},
+    {"G615", MaliGPU::kG615},
+    {"G710", MaliGPU::kG710},
+    {"G610", MaliGPU::kG610},
+    {"G510", MaliGPU::kG510},
+    {"G310", MaliGPU::kG310},
+    {"G78", MaliGPU::kG78},
+    {"G68", MaliGPU::kG68},
+    {"G77", MaliGPU::kG77},
+    {"G57", MaliGPU::kG57},
+
+    // Bifrost
+    {"G76", MaliGPU::kG76},
+    {"G72", MaliGPU::kG72},
+    {"G52", MaliGPU::kG52},
+    {"G71", MaliGPU::kG71},
+    {"G51", MaliGPU::kG51},
+    {"G31", MaliGPU::kG31},
+
+    // These might be Vulkan 1.0 Only.
+    {"T880", MaliGPU::kT880},
+    {"T860", MaliGPU::kT860},
+    {"T830", MaliGPU::kT830},
+    {"T820", MaliGPU::kT820},
+    {"T760", MaliGPU::kT760},
+};
+
+AdrenoGPU GetAdrenoVersion(std::string_view version) {
+  /// The format that Adreno names follow is "Adreno (TM) VERSION".
+  auto paren_pos = version.find(")");
+  if (paren_pos == std::string::npos) {
+    return AdrenoGPU::kUnknown;
+  }
+  auto version_string = version.substr(paren_pos + 2);
+  const auto& result = kAdrenoVersions.find(version_string);
+  if (result == kAdrenoVersions.end()) {
+    return AdrenoGPU::kUnknown;
+  }
+  return result->second;
+}
+
+MaliGPU GetMaliVersion(std::string_view version) {
+  // These names are usually Mali-VERSION or Mali-Version-EXTRA_CRAP.
+  auto dash_pos = version.find("-");
```

Comment:
Here I would search for 'Mali-' just to be as specific as possible. Are the version strings always at the start of `version`? You could also be super pedantic and check that it starts at position 0.

Could we also just use regex? 

```c++
  std::regex reg("^Mali-([\w]+)(-[\w]+)?");
  std::smatch match;
  if (std::regex_search(version, match, re) && match.size() > 1)  {
    const auto& result = kMaliVersions.find(match::str(1));
    if (result == kMaliVersions.end()) {
      return MaliGPU::kUnknown;
    }
  }
```

Reply:
e.g. https://regex101.com/r/Ae7yf8/1

Reply:
I'm terrified of c++ regex. Though the google c++ style guide says not to use them, that is only because they have a better alternative in RE2 (see https://engdoc.corp.google.com/eng/doc/devguide/cpp/styleguide.md?cl=head )

Reply:
I changed this to match more of the string.